### PR TITLE
Fire - Tram Bui

### PR DESCRIPTION
**Pull request by trambui09**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In the work model, I wrote a top_ten method that would select the top 10 works from each category ordered by most votes to least votes
Describe how you approached testing that model method. What edge cases did you come up with? |  Testing the top_ten method was tricky, I decided to use the fixtures to make 10 albums and a couple of books, then in the vote fixtures casted a couple of votes for different works. I tested if the top_ten return works in the right category and class. Then checked if it returned the work in the order of most votes to least votes. Also that it would return all the works in the category if there's less than 10 works. 
What are `session` and `flash`? What is the difference between them? | Sessions are like cookies, they keep track of which user is logged in and keep track of that user. flash is when you want to the message to persist to let the user know what happened. Whether what they did was successful or not
What was one thing that you gained more clarity on through this assignment? | creating custom methods and write appropriate tests for them 
What is the Heroku URL of your deployed application? | https://tram-media-ranker.herokuapp.com/

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? |

### CODE REVIEW

File: app/models/vote.rb [hunk 1]
```diff
@@ -0,0 +1,13 @@
+class Vote < ApplicationRecord
+  belongs_to :work
+  belongs_to :user
+
+  # validates :work uniqueness: { scope: :user }
+  # validates_uniqueness_of :work_id, scope: :user_id
+  # validates_uniqueness_of :vote, scope: [:work_id, :user_id]
+
+  validates :user, uniqueness: { scope: :work,
+                                    message: "has already voted for this work"
```
Comment (kaidamasaki) [inline app/models/vote.rb:10]: Nice use of a scope here!

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  def current_user
+    @current_user = User.find_by(id: session[:user_id])
+
+    # unless @current_user
+    #   # TODO: must be logged in to upvote
+    #   flash[:error] = "You must be logged in to see this page"
+    #   redirect_to root_path
+    #   return
+    # end
```
Comment (kaidamasaki) [inline app/controllers/application_controller.rb:13]: You have a lot of commented out code in this project.  

It makes it a little confusing to read.  In the future you should clean this up if you have time before you turn things in.

File: app/controllers/works_controller.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    # @work = Work.find_by(id: params[:id])
+    # @work.votes.create
+    # redirect_to works_path
+    user = User.find_by(id: session[:user_id])
+
+    if user.nil?
+      flash[:error] = "A problem occured: You must log in to do that"
+      redirect_to works_path
+      return
+    end
```
Comment (kaidamasaki) [inline app/controllers/works_controller.rb:105]: You had a `require_login` controller filter that already accomplished this but didn't use it.

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    return work_with_high_vote_id.last
+    # return Work.all.sort_by { |work | work.votes.count }.max
+  end
+
+  def self.top_ten(category)
+    # return Work.where(category: category).sample(10)
+    return Work.where(category: category)
+               .sort_by { |category_work| category_work.votes.count }
+               .reverse[0..9]
```
Comment (kaidamasaki) [inline app/models/work.rb:38]: You can use the [order](https://apidock.com/rails/ActiveRecord/QueryMethods/order) method to avoid having to reverse things and the [limit](https://apidock.com/rails/v5.2.3/ActiveRecord/QueryMethods/limit) method to tell the database you only want 10.

```suggestion
    return Work.where(category: category)
               .sort_by { |category_work| category_work.votes.count }
               .reverse[0..9]
```

File: app/views/homepages/_category.html.erb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+<section class="top-ten__list-container" >
+  <h2 class="top-ten__header">Top <%= category.capitalize %>s</h2>
+  <% works = Work.top_ten("#{category}")  %>
```
Comment (kaidamasaki) [inline app/views/homepages/_category.html.erb:3]: You should avoid calling model methods directly in views.

Instead you should set an instance variable in the appropriate controller.

File: app/views/layouts/application.html.erb [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+          </li>
+          <li class="nav-item app-header__nav_item">
+            <%= link_to "View all users", users_path, class: "nav-link" %>
+          </li>
+        </ul>
+
+        <ul class="nav app-header__user-nav-container">
+<!--          if there is a session id, present a log out buttong and button to take them to the user show page-->
+          <% if session[:user_id] %>
+            <% user = User.find_by(id: session[:user_id]) %>
```
Comment (kaidamasaki) [inline app/views/layouts/application.html.erb:39]: You should have used your filter for finding the current user instead of calling the model method and interacting with session directly here.

File: test/models/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  describe "relations" do
+    before do
+      new_user.save
+      # second_user = users(:tram)
+
+      @new_work = works(:cowboy)
+      @second_work = works(:undocumented)
+
+      vote_1 = Vote.create(user_id: new_user.id, work_id: @new_work.id)
+      vote_2 = Vote.create(user_id: new_user.id, work_id: @second_work.id)
```
Comment (kaidamasaki) [inline test/models/user_test.rb:20]: When creating data in `before` blocks I recommend using `create!` instead of `create` because it will raise an exception if something goes wrong instead of silently failing to save.

**Review by kaidamasaki — COMMENTED** [review]
Great job!  Everything functioned well and you had a lot of good attention to detail.

# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️?
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | Filter was unused.
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | ✔️
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | Missing validation tests.
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | Too much commented out code.
| Concise | ✅
| Logical/Organized | ✅
